### PR TITLE
Garbage Collection fka Nodes docs feedback 4

### PR DESCRIPTION
**Pull request by mburke5678**
Comments per @sunilcio 
https://bugzilla.redhat.com/show_bug.cgi?id=1688637#c11

### CODE REVIEW

**Comment by mburke5678** [comment]
Garbage collection questions still need to be resolved.

**Comment by mburke5678** [comment]
@sjenning PTAL

**Comment by mburke5678** [comment]
@sjenning Please advise on how to address @sunilcio comments above.

**Comment by mburke5678** [comment]
@sunilcio @sjenning I updated the garbage collection topics:
https://github.com/openshift/openshift-docs/pull/14453/commits/07511f60deeda2c4915eb942a2bf4b4053866b53

**Comment by sunilcio** [comment]
@mburke5678 are the updated garbage collection section published somewhere so that I can check and verify?

**Comment by mburke5678** [comment]
@sunilcio Preview file is here:
http://file.rdu.redhat.com/mburke/BZ-1688637-4/nodes/nodes/nodes-nodes-garbage-collection.html

Thank you for any feedback.

**Comment by sunilcio** [comment]
@mburke5678 thanks for the preview link, updated KubeletConfig yaml for container garbage collection is not getting applied. I have opened below BZ to track this request and to block our test cases related to container gc with reference to this BZ.

https://bugzilla.redhat.com/show_bug.cgi?id=1705901

I guess the syntax and format of `eviction soft` & `eviction soft grace period` options might be incorrect.

**Comment by mburke5678** [comment]
@sunilcio How do you know that the garbage collection YAML is not getting applied? What results did you expect?
@sjenning If you or someone on the Pod team can help with configuring garbage collection for 4.1, I would greatly appreciate it.

**Comment by sunilcio** [comment]
@mburke5678 when we create KubeletConfig CR, if it is valid a new rendered-worker-xxx will be created and can be checked with `oc get machineconfig` output and the custom kubelet config will be reflected on compute node's /etc/kubernetes/kubelet.conf file

**Comment by mburke5678** [comment]
@sunilcio Thank you. Just to be sure, I see a `rendered-worker-` machineconfiguration on a newly installed cluster. Will there be an a new `rendered-worker-` MC?

```
oc get machineconfig

NAME                                                        GENERATEDBYCONTROLLER      IGNITIONVERSION   CREATED
rendered-master-546383f80705b6663682c88d5aeaba93            4.1.0-201904211700-dirty   2.2.0             6h39m
rendered-worker-06f0c762bd8385af3a7f55017fea44e2            4.1.0-201904211700-dirty   2.2.0             6h39m

```

**Comment by sunilcio** [comment]
@mburke5678 yes, there will be a new rendered-worker MC. The one already present is around 6 hours old which I guess is the default one created during cluster creation. When we apply new KubeletConfig, a new rendered-worker will be created and in `oc get machineconfigpool` output we can see worker UPDATING status is true.

**Comment by sunilcio** [comment]
@mburke5678 as a working example of KubeletConfig, you can try setting `maxPods` & `podsPerCore` config documented at https://docs.openshift.com/container-platform/4.1/nodes/nodes/nodes-nodes-managing-max-pods.html#nodes-nodes-managing-max-pods-about-nodes-nodes-jobs

**Comment by mburke5678** [comment]
@sunilcio Have you discovered anything about how to configure garbage collection in 4.1?

**Comment by sunilcio** [comment]
@mburke5678 was not able to check this further. @sjenning can you please help here and check if we are setting the eviction parameter and format wrong?

**Comment by sjenning** [comment]
Regarding the deprecated flags (deprecated, not removed)

```
minimum-container-ttl-duration
maximum-dead-containers-per-container
maximum-dead-containers
```
The right thing has already been done, doc'ing the eviction options instead.

Image flags are not deprecated and should be documented.

This all LGTM

File: modules/nodes-nodes-viewing-listing-pods.adoc [hunk 1]
```diff
@@ -12,11 +12,25 @@ You can list all the pods on a specific node.
...
 +
 ----
-$ oc describe node <node1> <node2> --list-pods [--pod-selector=<pod_selector>] [-o json|yaml]
+$ oc describe node <node1> <node2>
+----
++
+For example:
++
+----
+$ oc describe node ip-10-0-128-218.ec2.interna
```
Comment (rphillips) [inline modules/nodes-nodes-viewing-listing-pods.adoc]: internal typo

File: modules/nodes-nodes-garbage-collection-configuring.adoc [hunk 2]
```diff
@@ -35,41 +35,109 @@ metadata:
...
-    ImageGCLowThresholdPercent: 80 <5>
-----
-<1> Assign a name to CR.
-<2> Specify the label to apply the configuration change.
-<3> Specify the minimum age for an unused image before it is garbage collected. A value of `0` means no limit.
-<4> Specify the percent of disk usage after which image garbage collection is always run.
-<5> Specify the percent of disk usage before which image garbage collection is never run.
-
-. Create the garbage collection object:
+  kubeletArguments:
```
Comment (rphillips) [inline modules/nodes-nodes-garbage-collection-configuring.adoc]: we don't have kubeletArguments anymore in 4.x

Comment (sjenning) [reply modules/nodes-nodes-garbage-collection-configuring.adoc]: good catch.  @mburke5678 just remove the `kubeletArguments:` line and we need to reformat the eviction stuff below as well.  replace `eviction-soft` with `EvictionSoft` and `eviction-soft-grace-period` with `EvictionSoftGracePeriod`.

File: modules/nodes-nodes-garbage-collection-configuring.adoc [hunk 3]
```diff
@@ -35,41 +35,110 @@ metadata:
...
+----
+apiVersion: machineconfiguration.openshift.io/v1
+kind: KubeletConfig
+metadata:
+  name: gc-image <1>
+spec:
+  machineConfigSelector:
+    matchLabels:
+      custom-kubelet: small-pods <2>
+  kubeletArguments:
```
Comment (rphillips) [inline modules/nodes-nodes-garbage-collection-configuring.adoc]: kubeletArguments here as well.

**Comment by rphillips** [comment]
I'm testing the instructions as well... I'll report back once I have gone through them.

File: modules/nodes-nodes-garbage-collection-configuring.adoc [hunk 4]
```diff
@@ -35,41 +35,110 @@ metadata:
...
+    eviction-soft-grace-period: <5>
+    - memory.available=1m30s
+    - nodefs.available=1m30s
+    - nodefs.inodesFree=1m30s
+    - imagefs.available=1m30s
+    - imagefs.inodesFree=1m30s
+----
+<1> Name for the object.
+<2> Selector label.
+<3> Type of eviction: `eviction-soft` or `eviction-hard`.
```
Comment (sjenning) [inline modules/nodes-nodes-garbage-collection-configuring.adoc]: This also needs to change to `EvictionSoft` and `EvictionHard`

File: modules/nodes-nodes-garbage-collection-configuring.adoc [hunk 5]
```diff
@@ -35,41 +35,110 @@ metadata:
...
+<4> Eviction thresholds based on a specific eviction trigger signal.
+<5> Grace periods for the soft eviction. This parameter does not apply to `eviction-hard`.
```
Comment (sjenning) [inline modules/nodes-nodes-garbage-collection-configuring.adoc]: same

File: modules/nodes-nodes-garbage-collection-configuring.adoc (hunk 2 above)
Comment (mburke5678) [reply modules/nodes-nodes-garbage-collection-configuring.adoc]: @rphillips @sjenning  Is this right?
```
spec:
  machineConfigSelector:
    matchLabels:
      custom-kubelet: small-pods
    EvictionSoft:
    - memory.available<100Mi
    - nodefs.available<10%
    - nodefs.inodesFree<5%
    - imagefs.available<15%
    - imagefs.inodesFree<10%
   EvictionSoftGracePeriod:
    - memory.available=1m30s
    - nodefs.available=1m30s
    - nodefs.inodesFree=1m30s
    - imagefs.available=1m30s
    - imagefs.inodesFree=1m30s
```

**Comment by mburke5678** [comment]
@sunilcio Please take a look at the updates from Ryan and Seth

File: modules/nodes-nodes-garbage-collection-configuring.adoc [hunk 6]
```diff
@@ -35,41 +35,107 @@ metadata:
...
+metadata:
+  name: gc-image <1>
+spec:
+  machineConfigSelector:
+    matchLabels:
+      custom-kubelet: small-pods <2>
+    ImageGCHighThresholdPercent: <3>
+    - "85"
+    ImageGCLowThresholdPercent: <4>
+    - "80"
```
Comment (rphillips) [inline modules/nodes-nodes-garbage-collection-configuring.adoc]: We need a kubeletConfig attribute in here.

```
spec:
  machineConfigSelector:
    matchLabels:
      custom-kubelet: small-pods <2>
  kubeletConfig:
    ImageGCHighThresholdPercent: <3>
    - "85"
    ImageGCLowThresholdPercent: <4>
    - "80"
```

Comment (rphillips) [reply modules/nodes-nodes-garbage-collection-configuring.adoc]: Updated to lowercase image.

File: modules/nodes-nodes-garbage-collection-images.adoc [hunk 7]
```diff
@@ -26,11 +26,14 @@ a Custom Resource.
...
-|`*image-gc-high-threshold*`
+|`*ImageMinimumGCAge*`
+|The minimum age for an unused image before the image is removed by garbage collection. The default is *2m*.
+
+|`*ImageGCHighThresholdPercent*`
 |The percent of disk usage (expressed as an integer) which triggers image
 garbage collection. The default is *85*.
 
-|`*image-gc-low-threshold*`
+|`*ImageGCLowThresholdPercent*`
```
Comment (rphillips) [inline modules/nodes-nodes-garbage-collection-images.adoc]: we may want to lowercase image here as well.

**Comment by rphillips** [comment]
Updated with a couple other changes. thank you @mburke5678

File: modules/nodes-nodes-garbage-collection-configuring.adoc (hunk 3 above)
Comment (weinliu) [reply modules/nodes-nodes-garbage-collection-configuring.adoc]: machineConfigSelector => machineConfigPoolSelector

[the remark this answers is not in the document]
Comment (weinliu) [reply modules/nodes-nodes-garbage-collection-configuring.adoc]: machineConfigSelector => machineConfigPoolSelector

**Comment by sunilcio** [comment]
@mburke5678 tested below config which works. Seth & Ryan helped with the new config!
For container gc:
~~~
  kubeletConfig:
    evictionSoft:
      memory.available: "500Mi"
      nodefs.available: "10%"
      nodefs.inodesFree: "5%"
      imagefs.available: "15%"
      imagefs.inodesFree: "10%"
    evictionSoftGracePeriod:
      memory.available: "1m30s"
      nodefs.available: "1m30s"
      nodefs.inodesFree: "1m30s"
      imagefs.available: "1m30s"
      imagefs.inodesFree: "1m30s"
    evictionHard:
      memory.available: "200Mi"
      nodefs.available: "5%"
      nodefs.inodesFree: "4%"
      imagefs.available: "10%"
      imagefs.inodesFree: "5%"
    evictionPressureTransitionPeriod: 0s
~~~
For image gc:
~~~
  kubeletConfig:
    imageMinimumGCAge: 5m
    imageGCHighThresholdPercent: 80 
    imageGCLowThresholdPercent: 75
~~~

**Comment by sunilcio** [comment]
@mburke5678 one more thing we need to also include in doc is that we we want to configure both image garbage collection and container garbage collection, then we need to combing parameters for both in single kubeletConfig. If we create 2 separate kubeletConfig, then the 2nd one will override 1st one. @rphillips is my understanding correct? 
Below is example of kubeletConfig including hard & soft eviction along with maxPods and other parameters.
~~~
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: gc-container
spec:
  machineConfigPoolSelector:
    matchLabels:
      custom-kubelet: small-pods 
  kubeletConfig:
    evictionSoft:
      memory.available: "500Mi"
      nodefs.available: "10%"
      nodefs.inodesFree: "5%"
      imagefs.available: "15%"
      imagefs.inodesFree: "10%"
    evictionSoftGracePeriod:
      memory.available: "1m30s"
      nodefs.available: "1m30s"
      nodefs.inodesFree: "1m30s"
      imagefs.available: "1m30s"
      imagefs.inodesFree: "1m30s"
    evictionHard:
      memory.available: "200Mi"
      nodefs.available: "5%"
      nodefs.inodesFree: "4%"
      imagefs.available: "10%"
      imagefs.inodesFree: "5%"
    evictionPressureTransitionPeriod: 0s
    imageMinimumGCAge: 5m
    imageGCHighThresholdPercent: 80 
    imageGCLowThresholdPercent: 75
    podsPerCore: 12 
    maxPods: 240
~~~

**Comment by rphillips** [comment]
That is correct. We only allow one kubelet config per machine config pool.

**Pull request merged**